### ISSUE TITLE
KeeAgent gets 'stuck' and wont provide keys

### ISSUE THREAD

**Issue by saxonww** [dlech/KeeAgent#109]
I've noticed that the agent will get stuck and stop providing keys. I am not sure if this is a KeeAgent problem or an SSH client issue; I am using KeeAgent w/ OpenSSH 6.8p1 via Cygwin (x64).

When this happens, I can open the KeePass options dialog (right click on tray icon -> options) and click OK, which will interrupt the stuck ssh attempt and present a password prompt. Subsequent login attempts will use the agent successfully until it gets stuck again.

**Comment by dlech** [comment]
This sounds like a difficult problem to reproduce. I won't be able to do anything unless I have a way to reproduce the issue. If you are up to it, you could try running it in Visual Studio and you should be able to figure out where it is getting stuck.

KeeAgent actually uses a network socket to communicate with ssh in Cygwin, so when you click OK in the Options dialog, it is closing the network socket and creating a new one, which is why it starts working again.

**Comment by kgraefe** [comment]
I have the same issue using Cygwin x86 on a x64 OS. My current workaround is to downgrade OpenSSH to 6.7.p1.

**Comment by phjr** [comment]
Same problem here. Works on 6.7p1 but not on 6.8p1

**Comment by kgraefe** [comment]
For me the key agent hangs exactly when I try to start a SSH connection while there is another connection open. It does not hang when I closed all SSH sessions before starting the next one. Plus if I close the first session when a second session hangs the second session becomes available.

Maybe OpenSSH 6.8p1 holds the socket open for the whole time it runs and 6.7p1 closes the socket as soon as it got the key?

**Comment by saxonww** [comment]
I'm not sure but I can confirm that downgrade to 6.7p1 fixed it for me. It
would be interesting to know if a Linux/BSD user with OpenSSH 6.8p1 and
Mono experiences the same issue. It might be a Cygwin problem.

**Comment by r2evans** [comment]
Not sure it helps with troubleshooting, but just in case ...

With ssh in triple-verbose, the connection stalls after being authenticated and "callback done":

```
...snip...
debug1: Authentications that can continue: publickey,password
debug3: start over, passed a different list publickey,password
debug3: preferred publickey,keyboard-interactive,password
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Offering RSA public key: /c/Users/r2/.ssh/id_rsa
debug3: send_pubkey_test
debug2: we sent a publickey packet, wait for reply
debug1: Server accepts key: pkalg ssh-rsa blen 279
debug2: input_userauth_pk_ok: fp SHA256:(blah blah blah)
debug3: sign_and_send_pubkey: RSA SHA256:(blah blah blah)
debug1: Enabling compression at level 6.
debug1: Authentication succeeded (publickey).
Authenticated to (my ip address).
debug1: channel 0: new [client-session]
debug3: ssh_session2_open: channel_new: 0
debug2: channel 0: send open
debug1: Requesting no-more-sessions@openssh.com
debug1: Entering interactive session.
debug2: callback start
```

It just hangs at that point. If I close keepass (File > Quit), it then continues with the connection:

```
debug1: ssh_get_authentication_socket: Connection refused
debug2: fd 3 setting TCP_NODELAY
debug3: ssh_packet_set_tos: set IP_TOS 0x10
debug2: client_session2_setup: id 0
debug2: channel 0: request pty-req confirm 1
debug2: channel 0: request shell confirm 1
debug2: callback done
debug2: channel 0: open confirm rwindow 0 rmax 32768
debug2: channel_input_status_confirm: type 99 id 0
debug2: PTY allocation request accepted on channel 0
debug2: channel 0: rcvd adjust 2097152
debug2: channel_input_status_confirm: type 99 id 0
debug2: shell request accepted on channel 0
```

However, KeePass is still showing in the task manager, easily killed but should have already been closed.

Failed under: OpenSSH_6.9p1, OpenSSL 1.0.2d 9 Jul 2015
Worked fine under: OpenSSH_6.7p1, OpenSSL 1.0.2d 9 Jul 2015
KeePass-2.29, KeeAgent-0.7.4.0

Thanks, Dave, I appreciate the work.

**Comment by dlech** [comment]
This should be fixed in v0.7.5 (just posted on my [website](http://lechnology.com/software/keeagent/)). 

Thanks for the hint about having two connections open at once. That was the key to be able to reproduce the problem. The tips about the possibility of sockets not closing and the KeePass.exe process hanging around after the UI was closed were helpful too.
